crates/key-value/src/lib.rs

```diff
@@ -138,6 +139,100 @@ impl key_value::HostStore for KeyValueDispatch {
     }
 }
 
+fn to_wasi_err(e: Error) -> wasi_keyvalue::store::Error {
+    match e {
+        Error::AccessDenied => wasi_keyvalue::store::Error::AccessDenied,
+        Error::NoSuchStore => wasi_keyvalue::store::Error::NoSuchStore,
+        Error::StoreTableFull => wasi_keyvalue::store::Error::Other("store table full".to_string()),
+        Error::Other(msg) => wasi_keyvalue::store::Error::Other(msg),
+    }
+}
+
+#[async_trait]
+impl wasi_keyvalue::store::Host for KeyValueDispatch {
+    async fn open(
+        &mut self,
+        identifier: String,
+    ) -> anyhow::Result<Result<Resource<wasi_keyvalue::store::Bucket>, wasi_keyvalue::store::Error>>
+    {
+        Ok(async {
+            if self.allowed_stores.contains(&identifier) {
+                let store = self
+                    .stores
+                    .push(self.manager.get(&identifier).await.map_err(to_wasi_err)?)
+                    .map_err(|()| {
+                        wasi_keyvalue::store::Error::Other("store table full".to_string())
+                    })?;
+                Ok(Resource::new_own(store))
+            } else {
+                Err(wasi_keyvalue::store::Error::AccessDenied)
+            }
+        }
+        .await)
+    }
+}
+
+use wasi_keyvalue::store::Bucket;
+#[async_trait]
+impl wasi_keyvalue::store::HostBucket for KeyValueDispatch {
+    async fn get(
+        &mut self,
+        self_: Resource<Bucket>,
+        key: String,
+    ) -> anyhow::Result<Result<Option<Vec<u8>>, wasi_keyvalue::store::Error>> {
+        let store = self.get_store(self_)?;
+        Ok(store.get(&key).await.map_err(to_wasi_err))
+    }
+
+    async fn set(
+        &mut self,
+        self_: Resource<Bucket>,
+        key: String,
+        value: Vec<u8>,
+    ) -> anyhow::Result<Result<(), wasi_keyvalue::store::Error>> {
+        let store = self.get_store(self_)?;
+        Ok(store.set(&key, &value).await.map_err(to_wasi_err))
+    }
+
+    async fn delete(
+        &mut self,
+        self_: Resource<Bucket>,
+        key: String,
+    ) -> anyhow::Result<Result<(), wasi_keyvalue::store::Error>> {
+        let store = self.get_store(self_)?;
+        Ok(store.delete(&key).await.map_err(to_wasi_err))
+    }
+
+    async fn exists(
+        &mut self,
+        self_: Resource<Bucket>,
+        key: String,
+    ) -> anyhow::Result<Result<bool, wasi_keyvalue::store::Error>> {
+        let store = self.get_store(self_)?;
+        Ok(store.exists(&key).await.map_err(to_wasi_err))
+    }
+
+    async fn list_keys(
+        &mut self,
+        self_: Resource<Bucket>,
+        cursor: Option<u64>,
+    ) -> anyhow::Result<Result<wasi_keyvalue::store::KeyResponse, wasi_keyvalue::store::Error>>
+    {
+        if cursor.unwrap_or_default() != 0 {
+            anyhow::bail!("list_keys: cursor not supported");
```

Comment:
Is the plan to eventually support the cursor?

Reply:
"Plan" is a strong word but, yes, it would be desirable to do on back-ends that support it.  For this initial implementation I stuck to using the existing KV implementation which unfortunately do not support cursors.